### PR TITLE
[RLlib] Experimental no-flatten option for actions/prev-actions.

### LINKED ISSUE #20589: [Bug] Tuple Action Space in Behavior Cloning
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I’m using this [saving experience example](https://github.com/ray-project/ray/blob/releases/1.8.0/rllib/examples/saving_experiences.py) to collect a dataset for Behavior Cloning (BC) algorithm. The action space of my environment is tuple. After feeding the saved experience into BC, I got the following error:

```
ValueError: The two structures don't have the same nested structure.
 
First structure: type=ndarray str=[[0 1]
[1 0]
[1 0]
[0 0]
[0 1]
[0 0]
[1 0]
[0 1]
[1 0]
[0 0]]
 
Second structure: type=tuple str=(Discrete(2), Discrete(2))
```
This error was trigged by `normalize_action()` at `ray/rllib/utils/spaces/space_utils.py`. 

### Versions / Dependencies

Ray 1.8.0
Python 3.6
MacOS Big Sur

### Reproduction script

An easy way to reproduce the error is to change the CartPole environment into a tuple action space. Specifically, the following steps will reproduce the error mentioned above:

1. Use the following action space
```
        self.action_space = spaces.Tuple((
            spaces.Discrete(2),
            spaces.Discrete(2)
        ))
```
to replace the original action space `self.action_space = spaces.Discrete(2)`

2. Use [saving experience example](https://github.com/ray-project/ray/blob/releases/1.8.0/rllib/examples/saving_experiences.py) to collect a small dataset

3. Use the following config for training:
```
bc-cartpole:
    env: CartPole-TupleAction
    run: BC
    local_dir: "~/workspace/temp"  # directory to store checkpoints
    checkpoint_at_end: False
    stop:
        timesteps_total: 10000
    config:
        # Works for both torch and tf.
        framework: torch
        num_workers: 0
        # actions_in_input_normalized: True
        input: "/home/ubuntu/ray_playground/tuple_action_bc/demo_data"
```

### Anything else

This error happens every time when training with BC on tuple action space environments. 

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by sven1977**
This PR introduces an experimental config flag `_disable_action_flattening` that can be set to True (default: False) for:
- leaving model-computed actions as-is inside the SampleCollectors' buffers.
- writing actions unflattened (possibly nested) to disk (JSONWriter).
- reading these unflattened (possibly nested) action structures from disk (JSONReader for offline RL).
- sending these unflattened (possibly nested) action structures as-is to a model that requires prev-actions inputs.

The near-future plan is to deprecate action flattening altogether and to always leave actions as-is for more transparency and user-friendliness, similar to our parallel effort to deprecate observation preprocessors.

A new test case has been added that checks, whether complex action spaces are handled correctly with both settings (_disable_action_flattening=True|False), with an output writer (does the SampleBatch written contain the correct action format?), and with a subsequent offline RL Trainer that reads from the previously generated json file and handles flattened/non-flattened actions well.

## Why are these changes needed?

Closes Issue #20589 

## Related issue number

Closes #20589 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/BUILD [hunk 1]
```diff
@@ -1486,6 +1486,14 @@ py_test(
     srcs = ["tests/test_multi_agent_pendulum.py"]
 )
 
+py_test(
```
Comment (sven1977) [inline rllib/BUILD:1489]: Added a test case for action flattening (default behavior) and non-flattening (new experimental behavior).

File: rllib/models/tf/tf_action_dist.py [hunk 2]
```diff
@@ -115,6 +115,9 @@ def __init__(self,
             for input_ in tf.split(inputs, input_lens, axis=1)
         ]
         self.action_space = action_space
+        if self.action_space is None:
```
Comment (sven1977) [inline rllib/models/tf/tf_action_dist.py:118]: Bug fix.

File: rllib/models/tf/tf_action_dist.py [hunk 3]
```diff
@@ -135,6 +138,8 @@ def logp(self, actions: TensorType) -> TensorType:
             if isinstance(self.action_space, gym.spaces.Box):
                 actions = tf.reshape(
                     actions, [-1, int(np.product(self.action_space.shape))])
+            elif isinstance(self.action_space, gym.spaces.MultiDiscrete):
```
Comment (sven1977) [inline rllib/models/tf/tf_action_dist.py:141]: Bug fix.

Comment (gjoliver) [reply rllib/models/tf/tf_action_dist.py:141]: how did you discover this bug?
can we make it a unit test for tf_action_dist?

File: rllib/offline/json_reader.py [hunk 4]
```diff
@@ -173,31 +176,45 @@ def _try_parse(self, line: str) -> Optional[SampleBatchType]:
...
+                        "into offline file with the "
+                        "`_disable_action_flattening=True` flag OR as already "
+                        "normalized (between -1.0 and 1.0) values. When "
+                        "reading already normalized action values, make sure "
+                        "to set `actions_in_input_normalized=True`.")
+                else:
+                    batch[SampleBatch.ACTIONS] = normalize_action(
+                        batch[SampleBatch.ACTIONS], pol.action_space_struct)
             else:
                 for pid, b in batch.policy_batches.items():
```
Comment (gjoliver) [inline rllib/offline/json_reader.py:219]: a quick question, how come when we going through policy_batches, we don't have to worry about the error case that the offline actions are already flattened like above?
should we use the same logics to process these 2 cases?

File: rllib/offline/json_reader.py [hunk 5]
```diff
@@ -235,30 +252,76 @@ def _next_file(self) -> FileType:
             path = random.choice(self.files)
         return self._try_open_file(path)
 
+    def _from_json(self, data: str) -> SampleBatchType:
+        if isinstance(data, bytes):  # smart_open S3 doesn't respect "r"
+            data = data.decode("utf-8")
+        json_data = json.loads(data)
```
Comment (gjoliver) [inline rllib/offline/json_reader.py:258]: offline training json file may be a few GBs big.
we are loading the whole thing in one shot here right?

File: rllib/offline/json_reader.py [hunk 6]
```diff
@@ -235,30 +252,76 @@ def _next_file(self) -> FileType:
...
+        if data_type == "SampleBatch":
+            if self.ioctx.worker is not None and \
+                    len(self.ioctx.worker.policy_map) != 1:
+                raise ValueError(
+                    "Found single-agent SampleBatch in input file, but our "
+                    "PolicyMap contains more than 1 policy!")
+            for k, v in json_data.items():
+                json_data[k] = unpack_if_needed(v)
+            if self.ioctx.worker is not None:
+                policy = next(iter(self.ioctx.worker.policy_map.values()))
```
Comment (gjoliver) [inline rllib/offline/json_reader.py:275]: this looks a little weird, why iter then next? is this just
self.ioctx.worker.policy_map.values()[0] ?

**Review by gjoliver — COMMENTED** [review]
sorry about the delay, took me a little while to read this.
just a few quick questions.

at a high level, I feel like things are scattered around a bit too much., and will be hard to update and maintain.
if we can have a couple of utils, like

```
flatten_obs_based_on_space(obs, space, config)
flatten_action_based_on_spece(action, space, config)
```

then, we can technically do all these with a single flag check.
but, I don't know, that may be a lot of refactoring work, so just a random thought.

Thanks.

File: rllib/offline/json_reader.py (hunk 5 above)
Comment (sven1977) [reply rllib/offline/json_reader.py:258]: No, `data` is always only one single line.

File: rllib/offline/json_reader.py (hunk 6 above)
Comment (sven1977) [reply rllib/offline/json_reader.py:275]:
```
>>> d = {"a": 0}
>>> d.values()[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'dict_values' object is not subscriptable
```

File: rllib/offline/json_reader.py (hunk 4 above)
Comment (sven1977) [reply rllib/offline/json_reader.py:219]: Great catch! Let me fix this. We should share the error message.

File: rllib/models/tf/tf_action_dist.py (hunk 3 above)
Comment (sven1977) [reply rllib/models/tf/tf_action_dist.py:141]: Yeah, great question. True, this should be tested more exhaustively on the unit test level (distribution tests).

File: rllib/offline/json_reader.py (hunk 6 above)
Comment (sven1977) [reply rllib/offline/json_reader.py:275]: list(values) would force the entirety of values to be placed in a list first, which may not be necessary. We just want to start iterating and stop after the first iter.

Comment (gjoliver) [reply rllib/offline/json_reader.py:275]: oh yeah, python3 dict.values() is a dict_value thing.
ok, my bad.

**Comment by sven1977** [comment]
I'm with you on that we should have a place for all flattening/unflattening utils. For tf/torch/numpy.

The thinking here was that we should get rid of flattening in either case (for both observations and actions) and leave everything always as-is coming from the env or models. We can probably deprecate lots of the existing tools for that very soon, once we make this new flag (`_disable_action_flattening`) True by default (follow up PR).

**Comment by gjoliver** [comment]
> I'm with you on that we should have a place for all flattening/unflattening utils. For tf/torch/numpy.
> 
> The thinking here was that we should get rid of flattening in either case (for both observations and actions) and leave everything always as-is coming from the env or models. We can probably deprecate lots of the existing tools for that very soon, once we make this new flag (`_disable_action_flattening`) True by default (follow up PR).

oh sure. that sounds awesome!! I feel like we are getting rid of a lot of things, which is awesome 😎

**Comment by sven1977** [comment]
"Best part is no part" :D

File: rllib/offline/json_reader.py (hunk 5 above)
Comment (gjoliver) [reply rllib/offline/json_reader.py:258]: I see.

File: rllib/offline/json_reader.py (hunk 4 above)
Comment (gjoliver) [reply rllib/offline/json_reader.py:219]: I can't seem to find where the diff is for this fix.
but I am gonna assume we handle this now.
thanks.

**Review by gjoliver — APPROVED** [review]
I really hope we get rid of the preprocessing all-together.
all these if statements are gonna be impossible to maintain.
let's keep a list of things that we will definitely get rid of from ray 2.0

File: rllib/offline/json_reader.py (hunk 4 above)
Comment (sven1977) [reply rllib/offline/json_reader.py:219]: Oh, sorry, you are right. Will do this now before merging. Thanks for surfacing this again!

**Pull request merged**